services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/AutoProvisioning.java

```diff
@@ -258,4 +271,21 @@ private static boolean isProvisionAsGatewayEnabledForTenant(final Tenant tenant,
     private static CredentialsResult<JsonObject> getCredentialsResult(final int status, final String message) {
         return CredentialsResult.from(status, new JsonObject().put(Constants.JSON_FIELD_DESCRIPTION, message));
     }
+
+    private static Optional<String> generateDeviceIdFromTemplateIfConfigured(final Tenant tenant,
+            final X509Certificate clientCertificate) {
+        final String issuerDN = clientCertificate.getIssuerX500Principal().getName(X500Principal.RFC2253);
+        final String subjectDN = clientCertificate.getSubjectX500Principal().getName(X500Principal.RFC2253);
+
+        return Optional.ofNullable(tenant.getAutoProvisioningDeviceIdTemplate(issuerDN))
+                .map(template -> template.replaceAll(QUOTED_PLACEHOLDER_SUBJECT_DN, subjectDN))
+                .map(template -> {
+                    if (template.contains(RegistryManagementConstants.PLACEHOLDER_SUBJECT_CN)) {
+                        return Optional.ofNullable(AuthenticationConstants.getCommonName(subjectDN))
+                                .map(cn -> template.replaceAll(QUOTED_PLACEHOLDER_SUBJECT_CN, cn))
+                                .orElseThrow(() -> new IllegalArgumentException("subject DN does not contain a CN"));
```

Comment:
I guess this will result in a 500 status code being sent back to the device, right? In case there is no client certificate included in the request, we return a 404 Not Found, which seems to be a better fit FMPOV. WDYT?

Reply:
Sounds good. Maybe you can also include an appropriate error message?

Reply:
I have updated with _404 Bad Request_ and an error message and pushed those changes.